src/Orleans/Streams/SimpleMessageStream/SimpleMessageStreamProducer.cs

```diff
@@ -86,7 +95,12 @@ internal class SimpleMessageStreamProducer<T> : IInternalAsyncBatchObserver<T>
             if (isDisposed) throw new ObjectDisposedException(string.Format("{0}-{1}", GetType(), "OnNextAsync"));
 
             if (!connectedToRendezvous)
+            {
+                // In order to avoid potential concurrency errors, synchronously copy the input before yielding the
+                // thread. DeliverItem below must also be take care to avoid yielding before copying for non-immutable objects.
+                item = (T)this.serializationManager.DeepCopy(item);
```

Comment:
this is just to handle the case found in #3034. So I think this is good. Just in case my words is confusing , I will state it very clear.